src/main/java/loqor/ait/tardis/TardisExterior.java

```diff
@@ -32,34 +46,163 @@ public class TardisExterior extends TardisComponent {
-                WorldOps.updateIfOnServer(server.getWorld(tardis.travel().position().getWorld().getRegistryKey()),
-                        tardis.travel().position().getPos());
-                if (variantChange) {
-                    tardis.getExterior().setVariant(schema);
-                    WorldOps.updateIfOnServer(server.getWorld(tardis.travel().position().getWorld().getRegistryKey()),
-                            tardis.travel().position().getPos());
-                }
-                TardisEvents.EXTERIOR_CHANGE.invoker().onChange(tardis);
-            });
+            Identifier variantValue = buf.readIdentifier();
+
+            ExteriorVariantSchema schema = ExteriorVariantRegistry.getInstance()
+                    .get(variantValue);
+
+            if (!tardis.getExterior().update(schema, variantChange))
+                return;
+
+            server.execute(() -> StackUtil.playBreak(player));
+        }));
+
+        TardisEvents.ENTER_FLIGHT.register(tardis -> {
+            if (isDisguised(tardis))
+                tardis.getExterior().clearDisguise();
+        });
+
+        TardisEvents.LANDED.register(tardis -> {
+            if (isDisguised(tardis))
+                tardis.getExterior().applyDisguise();
         });
+
+        TardisEvents.SEND_TARDIS.register((tardis, player) -> {
+            if (isDisguised(tardis) && !player.isInTeleportationState())
+                tardis.getExterior().applyDisguise(player);
+        });
+
+        TardisEvents.EXTERIOR_CHANGE.register(tardis -> {
+            TardisExterior exterior = tardis.getExterior();
+
+            if (exterior.getVariant() instanceof AdaptiveVariant) {
+                exterior.applyDisguise();
+                return;
+            }
+
+            exterior.clearDisguise();
+        });
+
+        TardisEvents.USE_DOOR.register((tardis, player) -> {
+            if (player == null)
+                return;
+
+            TardisExterior exterior = tardis.getExterior();
+
+            if (!(exterior.getVariant() instanceof AdaptiveVariant))
+                return;
+
+            if (tardis.door().isClosed()) {
+                exterior.applyDisguise(player);
+                return;
+            }
+
+            DirectedGlobalPos.Cached cached = tardis.travel().position();
+            Optional<ExteriorBlockEntity> blockEntity = exterior.findExteriorBlock();
+
+            if (blockEntity.isEmpty())
+                return;
+
+            player.networkHandler.sendPacket(new BlockUpdateS2CPacket(cached.getWorld(), cached.getPos()));
+            player.networkHandler.sendPacket(new BlockUpdateS2CPacket(cached.getWorld(), cached.getPos().up()));
+            player.networkHandler.sendPacket(BlockEntityUpdateS2CPacket.create(blockEntity.get()));
+        });
+
+        FakeBlockEvents.CHECK.register((player, state, pos) -> {
+            shitParticles(player.getServerWorld(), pos);
```

Comment:
lol i remember loqor getting upset at me cus of swears and non professionalism

Reply:
Lmaooo
Nahh this is peak code wdym